Refactor server API implementation

npalaska opened the pull request:
This PR attempts to refactor some of the server API implementations

- Remove global variable `app` 
- Keep `create_app` functionality clean and high level.
- Modularize the API implementations
  - Create Separate api module for upload, elasticsearch and graphql
- Remove custom implementation of CORS 
  - Use `flask_cors` module to add  CORS decorator
- Use the loaded PbenchServerConfig object instead of reassigning them to app.config

dbutenhof commented on lib/pbench/server/api/config.py:
```diff
@@ -0,0 +1,109 @@
+    DEBUG = False
+    TESTING = False
+
+    @property
+    def ELASTICSEARCH_URI(self):
+        return f"http://{self.CONFIG.conf['elasticsearch'].get('host')}:{self.CONFIG.conf['elasticsearch'].get('port')}"
+
+    @property
+    def GRAPHQL_URI(self):
+        return f"http://{self.CONFIG.conf['graphql'].get('host')}:{self.CONFIG.conf['graphql'].get('port')}"
+
+    # Add prod specific db config
```
For the others, it's arguable whether it makes more sense to pre-fetch and return local properties vs grabbing them from the composed config object. In these, however, I think that returning local properties instead of doing the string formatting on each call would make more sense.

npalaska replied:
yeah I could define a property to get elasticsearch_host and elasticsearch_port separately but then where do we need these properties as separate things, each time we ask for es host and port its because we want to form a URI so I thought I would just return the URI because thats what most of the time we need

dbutenhof replied:
No, what I meant is that it'd be (slightly) more efficient and a bit cleaner to construct this string in `__init__` and simply return the constructed property here rather than re-fetching the config options and rebuilding the template each time.

dbutenhof commented on lib/pbench/server/api/__init__.py:
```diff
@@ -48,59 +49,16 @@ def register_endpoints(api, app):
 
 
-def create_app():
+def create_app(server_config=ProdServerConfig):
```
I'd rather see you pass in an _object_ rather than a _class_ that gets instantiated here. That just feels awkward; and passing an object will provide more flexibility for testing, too.

npalaska replied:
you mean something like `create_app(server_config=ProdServerConfig()):` but then what do you mean by instantiating it? The config classes ideally wouldn't have any methods just constant variables

portante reviewed (changes requested):
Can you update the description for this PR with more details on what problem this is solving?

It appears to be creating a object to which properties are derived from the loaded configuration.  Can't we just add those properties to the existing `PbenchServerConfig` class?

dbutenhof commented on lib/pbench/server/api/__init__.py:
```diff
@@ -48,59 +49,16 @@ def register_endpoints(api, app):
-    app.config["REST_URI"] = app.config_server.get("rest_uri")
-    app.config["BIND_HOST"] = app.config_server.get("bind_host")
-    app.config["WORKERS"] = app.config_server.get("workers")
+    app.config.from_object(server_config())
```
`server_config()` here is instantiating an object of your class; I just think that constructor should be called from the client, passing in the object rather than a reference to the class.

portante commented on lib/pbench/server/api/__init__.py:
```diff
@@ -63,44 +57,35 @@ def create_app():
 
-    app = Flask(__name__)
-    api = Api(app)
+    app.config["SERVER_CONFIG"] = server_config.server_config
```
Why are we going through the trouble of still using `app.config`?  Why not just reference the `server_config` instead?

npalaska replied:
I was following the Flask config handling https://flask.palletsprojects.com/en/1.1.x/config/ where it says 

> Independent of how you load your config, there is a config object available which holds the loaded configuration values: The config attribute of the Flask object. This is the place where Flask itself puts certain configuration values and also where extensions can put their configuration values. But this is also where you can have your own configuration.

Flask `Config` is a subclass of a dictionary which comes with some pre-configured values and if we use extensions such as for example `flask_bcrypt` then we would need a `SECRET_KEY` or some key to bcrypt the password hash, this key can be stored as a config `key` (now it can also be just retrieved with PbenchServerConfig class, however some extensions can read the default value if the key is set). The advantage I see is that we dont have to pass the entire server_config object to other modules to get certain values (we just pass the app instance) and also for the testing when we import the `create_app` to make the `app_client` we can assign the `app_client.config = app.config` and therefore we so we dont need to make another serverconfig object

npalaska replied:
but on the second thought I agree it adds an extra redundancy for our use case

npalaska commented:
I have updated the PR description and name, since its now doing more than just app config changes. I took this PR as an opportunity to modularize the api implementation and some cleanups as mentioned in the description. I cherry-picked these changes (specifically modularize commit) from pbench_user implementation PR and added here since these changes shouldn't be part of that PR.

portante commented on lib/pbench/server/__init__.py:
```diff
@@ -143,6 +117,101 @@ def mocked_time():
+        return self.conf["pbench-server"]
+
+    @property
+    def logger(self):
```
This seems to be adding a new `logger` attribute (property) to a `PbenchServerConfig` class.  All the existing callers create their own loggers from the server config.  Can we have the API instance do the same?

npalaska replied:
Addressed.

portante commented on lib/pbench/server/api/resources/graphql_api.py:
```diff
@@ -0,0 +1,50 @@
+        except Exception:
+            self.logger.exception("Exception occurred GraphQL response construction")
+            abort(
+                500, message="INTERNAL ERROR",
```
I wonder why this formats differently from line 38?

npalaska replied:
Ahh I thinks its because of the `,` at the end, good catch

portante commented on lib/pbench/server/api/resources/upload_api.py:
```diff
@@ -0,0 +1,192 @@
+class HostInfo(Resource):
+    def __init__(self, server_config, logger, api):
+        self.server_config = server_config
+        self.api = api
```
It does not appear that `api` is actually used by this class.

portante commented on lib/pbench/server/api/resources/upload_api.py:
```diff
@@ -0,0 +1,192 @@
+            self.logger.exception(
+                "There was something wrong constructing the host info"
+            )
+            abort(500, message="There was something wrong with your request")
```
Feel free to create an issue since this is not really your code changes ...

Perhaps just report back to the client, "`INTERNAL ERROR`"?

portante commented on lib/pbench/server/api/resources/upload_api.py:
```diff
@@ -0,0 +1,192 @@
+            response = jsonify(
+                dict(
+                    message=f"{self.server_config.get('user')}@{self.server_config.get('host')}"
+                    f":{self.server_config.get('pbench-receive-dir-prefix')}-002"
```
Instead of fetching this each time, and getting the configuration error each time, perhaps we could fetch these values one in the `HostInfo` constructor, and raising the exception there if they are missing, and then just reference the constructed response.

portante commented on lib/pbench/server/api/resources/upload_api.py:
```diff
@@ -0,0 +1,192 @@
+class Upload(Resource):
+    def __init__(self, upload_directory, max_content_length, logger, api):
+        self.upload_directory = upload_directory
+        self.max_content_length = max_content_length
```
This item is specific to the `Upload` class, so perhaps we should pass the configuration object so that we don't need to endow knowledge of what the `Upload` class needs outside of the class itself.

portante commented on lib/pbench/server/api/resources/upload_api.py:
```diff
@@ -0,0 +1,192 @@
+        self.upload_directory = upload_directory
+        self.max_content_length = max_content_length
+        self.logger = logger
+        self.api = api
```
Not sure where `api` is used.

portante commented on lib/pbench/server/api/resources/elasticsearch_api.py:
```diff
@@ -0,0 +1,81 @@
+
+    def __init__(self, elasticsearch_uri, logger, api):
+        self.elasticsearch = elasticsearch_uri
+        self.api = api
```
Not sure where `api` is used.

portante commented on lib/pbench/server/api/resources/elasticsearch_api.py:
```diff
@@ -0,0 +1,81 @@
+    """Elasticsearch API for post request via server."""
+
+    def __init__(self, elasticsearch_uri, logger, api):
+        self.elasticsearch = elasticsearch_uri
```
This looks like a specific configuration that the `Elasticsearch` class could pull out of the configuration object so that knowledge of this class remains with the class.

npalaska replied:
yeah I agree, I did that before where I was passing whole config object and elasticsearch class will fetch the required parameters but then I changed it to something like give me only what I want framework, but you put a good point that this is a specific configuration for the Elasticsearch class and let it fetch what it wants

portante commented on lib/pbench/server/api/__init__.py:
```diff
@@ -63,361 +60,24 @@ def create_app():
+def create_app():
+    """Create Flask app with defined resource endpoints."""
+
     app = Flask(__name__)
```
Is `__name__` going to resolve to "`api`" or to "`__init__`"?  If the former, great; if the latter, then let's not use `__name__` but perhaps `api-server`, or something like that.

npalaska replied:
I am not 100% sure but I think it will resolve to the module name. ref: https://flask.palletsprojects.com/en/1.1.x/tutorial/factory/ so in our case `pbench.server.api`

> `__name__` is the name of the current Python module. The app needs to know where it’s located to set up some paths, and `__name__` is a convenient way to tell it that.

portante replied:
Try this out:

```
$ cat this.py
class Foo:
    def __init__(self):
        self.myname = "foo"

    def bar(self):
        print(self.myname, __name__)


def this():
    print(__name__)

print(__name__)
this()
Foo().bar()
$ python3 this.py
__main__
__main__
foo __main__
```

npalaska replied:
the yes but we dont directly execute the `__init__` file, we execute the file via gunicorn. to do that we import `create_app` in `shell.py` file `from pbench.server.api import create_app` and if I execute shell.py file I get `pbench.server.api` when I do `print(__name__)` inside `create_app()`

portante replied:
Right, so that is the module name, but we want a logger for the instance of the service.  I am not sure we should name the logger based on the module.  At least, in other parts of the server code, we name the logger explicitly, see https://github.com/distributed-system-analysis/pbench/blob/master/lib/pbench/server/indexer.py#L4415.

npalaska replied:
Ahh I see, thanks for the clarification

portante commented on lib/pbench/server/__init__.py:
```diff
@@ -143,6 +117,97 @@ def mocked_time():
+            return None
+
+    @property
+    def upload_directory(self):
```
Can we move this to the `Upload` constructor class instead?

portante commented on lib/pbench/server/__init__.py:
```diff
@@ -143,6 +117,97 @@ def mocked_time():
+            return ""
+
+    @property
+    def graphql(self):
```
Can we move this to the `GraphQL` class constructor instead?

portante commented on lib/pbench/server/__init__.py:
```diff
@@ -143,6 +117,97 @@ def mocked_time():
+        return self.conf["pbench-server"]
+
+    @property
+    def elasticsearch(self):
```
Can we move this to the `Elasticsearch` constructor instead?

portante commented on lib/pbench/server/__init__.py:
```diff
@@ -143,6 +117,97 @@ def mocked_time():
+        return self._get_conf("pbench-server", "rest_version")
+
+    @property
+    def max_content_length(self):
```
Can we move this to the `Upload` constructor instead?

portante reviewed (changes requested):
Not sure about all these new properties, as they appear to be very specific to one case, used once, and not general enough to be useful as properties.

npalaska commented on lib/pbench/server/api/resources/graphql_api.py:
```diff
@@ -0,0 +1,48 @@
+
+        if not json_data:
+            self.logger.warning("GraphQL: Invalid json object {}", request.url)
+            abort(400, message="GraphQL: Invalid json object in request")
```
I tried to address it here

portante commented on lib/pbench/cli/server/shell.py:
```diff
@@ -1,14 +1,18 @@
 
 
+server_config = get_server_config()
+application = create_app(server_config)
```
Seems like these should not be global variables but local only to `main()`.

npalaska replied:
I had to do this because the `gunicorn` expect the format to be `{module}:flask_app_instance` to run the app. If I define it inside the main function then I have to do something like `pbench.cli.server.shell:main()` where `main` function need to return the flask_app instance. Another alternative would be to create another file say `wsgi.py` to hold the application and `shell.py` would import it and then gunicorn can call `pbench.cli.server.wsgi:application()` where application function in `wsgi.py` would return the app instance

portante replied:
Right, but why not declare only `application` as `global`, but set its value to `None`.  Then in `main()`, fetch the server configuration, create the application instance, assigning it to the `application` global variable, and then invoke `gunicorn`?

This prevents the application and configuration from being loaded at module load time, so that we can unit test on demand, clearing out the value of the global variable only.

npalaska replied:
Actually while testing now, I realized this would break. We run the gunicorn using the python subprocess and thus it create a separate gunicorn process and for that process the `application` would still remain `None` because it wont pick the same instance of application that we tried to set in the `main`

portante commented on lib/pbench/server/api/resources/upload_api.py:
```diff
@@ -190,3 +194,25 @@ def allowed_file(filename):
+            __name__, "pbench-server", "pbench-receive-dir-prefix", self.logger
+        )
+        try:
+            return Path(f"{prdp}-002").resolve(strict=True)
```
Since we are returning a `Path` object here, line 106 can be changed to `path = self.upload_directory / controller`.

portante commented on lib/pbench/test/unit/server/conftest.py:
```diff
@@ -83,8 +83,9 @@ def client(pytestconfig, monkeypatch):
 
     from pbench.server.api import create_app, get_server_config
```
Shouldn't this be at the top of the file with all the other imports?

npalaska replied:
ahh yeah

dbutenhof commented on lib/pbench/server/api/__init__.py:
```diff
@@ -6,53 +6,43 @@
 
-    global app
 
+def get_server_config():
```
Seems like this should probably be in the shell, and simply passed in the config object; no need for shell to import this from the api just to construct the object and pass it back. There's nothing in here that's specific to the API implementation.

npalaska replied:
That's correct but we also need to import this function in `conftest.py` to build the flask client app for pytests and thats why I kept it here. (I can probably put it in the utils file but little skeptical about having to import `PbenchConfigClass` in the utils file coz in the future as we add more stuff and write utility functions we might encounter circular import scenarios. Correct me if I am wrong but util files shouldnt import important classes right?)

portante replied:
If we can avoid it, in principle, we don't want to modify the code just to accommodate tests.  Let's have the test environment accommodate the code.

dbutenhof commented on lib/pbench/server/api/resources/elasticsearch_api.py:
```diff
@@ -0,0 +1,82 @@
+                self.logger.debug(
+                    "No payload found in Elasticsearch post request json data"
+                )
+                es_response = requests.get(url)
```
You need to add the `es_response.raise_for_status()` here to activate the exception transformation. That's in my PR; however since you've moved everything your merge is going to be messy and it could easily be overlooked!

npalaska replied:
Thanks for the heads up

dbutenhof commented on lib/pbench/server/api/resources/graphql_api.py:
```diff
@@ -0,0 +1,51 @@
+        try:
+            # query GraphQL
+            gql_response = requests.post(self.graphql, json=json_data)
+        except requests.exceptions.ConnectionError:
```
`qql_response.raise_for_status()`

dbutenhof commented on lib/pbench/test/unit/server/conftest.py:
```diff
@@ -12,13 +11,13 @@
-[elasticsearch]
-host = elasticsearch.example.com
-port = 7080
+[elasticsearch]	
```
Did you mean to add trailing whitespace on these lines? How'd it get there?

npalaska replied:
Ahh yeah I remove them, I was playing around with it and it got added

dbutenhof reviewed (commented):
And of course you need to rebase with my changes, which is going to be enormous fun. :-)

npalaska commented:
> And of course you need to rebase with my changes, which is going to be enormous fun. :-)

I just rebased and indeed it is going to be lot of fun

npalaska commented on lib/pbench/server/__init__.py:
```diff
@@ -143,6 +115,55 @@ def mocked_time():
+    def server_config(self):
+        return self.conf["pbench-server"]
+
+    @property
+    def es_url(self):
+        try:
+            return f"http://{self.conf.get('elasticsearch', 'host')}:{self.conf.get('elasticsearch', 'port')}"
+        except (NoOptionError, NoSectionError):
+            return ""
+
+    @property
+    def index_prefix(self):
+        try:
+            return self.conf.get("Indexing", "index_prefix")
+        except (NoOptionError, NoSectionError):
+            return ""
+
```
I believe we might need these properties as we add more es queries

portante replied:
The question is would we have code written in another module that talks to Elasticsearch NOT use the `elasticsearch_api` module?

The query_controllers module also talks to Elasticsearch, but why not make one module which handles the communication for both?

portante commented on lib/pbench/server/__init__.py:
```diff
@@ -143,6 +115,55 @@ def mocked_time():
+        return self.conf["pbench-server"]
+
+    @property
+    def es_url(self):
```
These seems pretty specific to the Elasticsearch module we have.

portante reviewed (changes requested):
Getting there, I think we can resolve the need for global state fairly easily, and address the separation of concerns with respect to the configuration properties and be ready to merge.

portante commented on lib/pbench/server/api/__init__.py:
```diff
@@ -12,11 +12,11 @@
-from pbench.server.api.resources.query_controllers import QueryControllers
 from pbench.common.logger import get_pbench_logger
+from pbench.server.api.resources.query_apis.elasticsearch_api import Elasticsearch
+from pbench.server.api.resources.query_apis.query_controllers import QueryControllers
```
Any particular reason for this change?

npalaska replied:
I just thought we could keep the elasticsearch related query apis in one place so its organized and easy to remove and/or add more apis or make changes that would reflect on all the es apis

portante replied:
Typically we have been following the PEP 8 on [Imports](https://www.python.org/dev/peps/pep-0008/#imports):

> Imports are always put at the top of the file, just after any module comments and docstrings, and before module globals and constants.
>
> Imports should be grouped in the following order:
>
> * Standard library imports.
> * Related third party imports.
> * Local application/library specific imports.
> You should put a blank line between each group of imports.

portante reviewed (changes requested):
Actually, needs a rebase ...

portante commented on lib/pbench/server/api/__init__.py:
```diff
@@ -72,375 +61,23 @@ def create_app():
         sys.exit(1)
```
If we are going to turn this into a utility function, we want to remove the calls to `sys.exit()`, and raise an exception that the outer layer can call `sys.exit()`.  See also #1955.

npalaska replied:
yeah good point, will do

portante reviewed (changes requested):
Let's address these `sys.exit()` move.  It was fine in `create_app()` before because that was the outer later, but burying `sys.exit()` calls can be tricky to deal with long term.

dbutenhof commented on lib/pbench/test/unit/server/conftest.py:
```diff
@@ -83,12 +82,22 @@ def teardown():
+
+
+@pytest.fixture
+def client(pytestconfig, monkeypatch, server_config):
```
Since you refactored the only `monkeypatch` reference into the `server_config` fixture, I don't think you need `monkeypatch` here anymore. Nor probably `pytestconfig` for that matter...

npalaska replied:
yes you are right its redundant

dbutenhof commented on lib/pbench/test/unit/server/test_requests.py:
```diff
@@ -28,36 +29,32 @@ def test_host_info(client, pytestconfig, caplog):
 
 class TestElasticsearch:
```
We should consider breaking these test classes off into separate files associated with each of the new resource classes.

(I'm not really pointing a finger at you, since I was the first to break off the new `QueryControllers` and I was lazy about this... but we should still think about doing it!)

npalaska replied:
yep I was thinking about that, will just go ahead then

dbutenhof replied:
Good start; we can unbundle the rest later. (And I suspect we'll want to generalize the `query_helper` fixture to mock behavior on the GraphQL and Elasticsearch pass-through APIs as well.)

portante commented on lib/pbench/test/unit/server/conftest.py:
```diff
@@ -2,8 +2,7 @@
 from pathlib import Path
-
-from pbench.server.api import create_app
+from pbench.server.api import create_app, get_server_config
```
Not something to change here, but PEP 8 specifies that the groups of the different "kinds" of imports (sorted) are separated by a space.  But since `black` and `flake8` don't really catch that we can live with it.

So really the imports should look like:

```
import pytest
import tempfile

from pathlib import Path

from pbench.server.api import create_app, get_server_config
```

npalaska replied:
I see, got it

dbutenhof commented on lib/pbench/server/api/resources/query_apis/__init__.py:
```diff
@@ -0,0 +1,33 @@
+from configparser import NoSectionError, NoOptionError
+
+
+def get_es_host(config):
+    try:
+        return config.get("elasticsearch", "host")
+    except (NoOptionError, NoSectionError):
+        return ""
+
+
+def get_es_port(config):
+    try:
+        return config.get("elasticsearch", "port")
+    except (NoOptionError, NoSectionError):
+        return ""
```
Are these actually used anywhere other than `get_es_url`?? It doesn't seem to me that they ought to be; in which case when don't we just elevate the URL to an `@property` on the config object in the first place and skip all this?

npalaska replied:
These are elasticsearch properties and are very specific to the Elasticsearch modules we have and thus I removed it from `PbenchServerConfig` `@properties` and added it here as a common helper function that can be used by number of future query modules. `es_host` and `es_port` are not used independently anywhere else (at least yet) but having a separate function helps to know whether the port information is missing or the host information from the config file.

portante reviewed (changes requested):
Needs a rebase ...

portante commented on lib/pbench/cli/server/shell.py:
```diff
@@ -1,14 +1,34 @@
+
+
+try:
+    server_config = get_server_config()
```
Do they really need to be the same server config object?

Why not have:

```
def app():
    return create_app(get_server_config())

def main():
    server_config = get_server_config()
    try:
        ...
```

npalaska replied:
In that case we would be creating two server config object instances where one instance is used just for getting gunicorn host port and workers information, is it fine?

portante replied:
Personally, I'd rather that than the global variable.  That global variable is created a module load time, which is not ideal and creates an inflexible interface.

portante reviewed (changes requested):
Looks good!  Let's rebase and we can merge.
